### PR TITLE
Fix for #PyDev-1055: Pydev does not recognize named unicode character escape in f-string

### PR DESCRIPTION
**Pull request by oluiscabral**
I created a new factory class to handle f-strings and check if it has a scape unicode character before code f-string variables definition.

I also ran JavaCC build.xml, so some code changes were automatic generated

### CODE REVIEW

File: plugins/org.python.pydev.parser/src/org/python/pydev/parser/grammar_fstrings/FStringsGrammarFactory.java [hunk 1]
```diff
@@ -0,0 +1,25 @@
+package org.python.pydev.parser.grammar_fstrings;
+
+import org.python.pydev.parser.jython.FastCharStream;
+
+public class FStringsGrammarFactory {
+
+    public static FStringsGrammar createGrammar(String s) {
+
+        while (s.contains("\\N{")) {
```
Comment (fabioz) [inline plugins/org.python.pydev.parser/src/org/python/pydev/parser/grammar_fstrings/FStringsGrammarFactory.java:9]: You could do something as:

```
int i;
while((i=s.indexOf("\\N{")) != -1) {
```

to avoid having to iterate more than once (one for `contains` and another one for the `indexOf`). Still, there are some issues in the code... the code below doesn't check if `s_.indexOf("}");` returns `-1` (so, if you have`\\N{` without closing it you could have issues) and the `s = s.replace(s.charAt(pos), '_');` that's done will actually replace all occurrences of `{`, which is not what you want (so, if your input has multiple `\\N` with code that should be  replaced as `{var}`, it won't find the `var` (please create test cases for these situations)).

Also, I don't think you need to create that many temporary strings, a single loop could probably do what you want... Probably something as the code below would be a bit better (note: I haven't run it, so, please update the code accordingly if something is wrong / tests fail -- it's mostly to provide the idea for the code).

```
        char[] c = s.toCharArray();
        for (int i = 0; i < c.length; i++) {
            if (c[i] == '\\' && i + 2 < c.length && c[i + 1] == 'N' && c[i + 2] == '{') {
                c[i + 2] = '_';
                i += 2;
                for (; i < c.length; i++) {
                    if (c[i] == '}') {
                        c[i] = '_';
                        break;
                    }
                }
            }
        }
```

You should also test in conditions which could have corner cases, such as just `\N` or `\N{` or `\N{\N{}`.

File: .gitignore [hunk 2]
```diff
@@ -53,4 +53,5 @@ plugins/org.python.pydev.core/pysrc/.project
 plugins/org.python.pydev.core/pysrc/.settings
 **/v2_indexcache/**
 plugins/com.python.pydev.runalltests/fake_homedir*
-.cache
\ No newline at end of file
+.cache
+plugins/org.python.pydev.parser/src/org/python/pydev/parser/grammar_fstrings/build.xml
```
Comment (fabioz) [inline .gitignore:57]: This file shouldn't be added to `.gitignore`.

File: plugins/org.python.pydev.parser/src/org/python/pydev/parser/grammar_fstrings/FStringsGrammarFactory.java [hunk 3]
```diff
@@ -0,0 +1,25 @@
+package org.python.pydev.parser.grammar_fstrings;
+
+import org.python.pydev.parser.jython.FastCharStream;
+
+public class FStringsGrammarFactory {
+
+    public static FStringsGrammar createGrammar(String s) {
```
Comment (fabioz) [inline plugins/org.python.pydev.parser/src/org/python/pydev/parser/grammar_fstrings/FStringsGrammarFactory.java:7]: You could have a different method and test just the string replacement directly (so, create a `public static char[] replaceNamedUnicodeVars(String s)` which will return a `char[]` with the return and test that directly to see input/output (instead of just testing checking the vars as it's now).

And this method would call that one when creating the grammar ;)

File: plugins/org.python.pydev.parser/src/org/python/pydev/parser/grammar_fstrings/build.xml [hunk 4]
```diff
@@ -1,7 +1,7 @@
 <project name="jython-parser" default="parser">
 
-    <property name="javaccHome" value="c:\bin\javacc-5.0" />
-    <property name="python" value="c:\bin\Miniconda\envs\py27_tests\python.exe" />
+    <property name="javaccHome" value="C:\Users\olccjr\Documents\Javacc\javacc-5.0" />
```
Comment (fabioz) [inline plugins/org.python.pydev.parser/src/org/python/pydev/parser/grammar_fstrings/build.xml:3]: I think that we could use env variables for that (so, please update the build to use environment variables instead of hardcoding -- please add comments to say what are the expected values for the environment variables for other users).

File: plugins/org.python.pydev.parser/tests/org/python/pydev/parser/FStringsParserTest.java [hunk 5]
```diff
@@ -76,7 +75,24 @@ private void checkError(String str, String... expected) throws ParseException {
         }
     }
 
+    private Tuple<FStringsAST, List> checkStr(String str, Set<String> exprs)
```
Comment (fabioz) [inline plugins/org.python.pydev.parser/tests/org/python/pydev/parser/FStringsParserTest.java:78]: I don't think you need this method.

File: plugins/org.python.pydev.parser/tests/org/python/pydev/parser/FStringsParserTest.java [hunk 6]
```diff
@@ -76,7 +75,24 @@ private void checkError(String str, String... expected) throws ParseException {
...
+            String contents = b.getContentsFromString(doc);
+            found.add(contents);
+        }
+
+        assertEquals(exprs.toString(), found.toString());
+        return ret;
+    }
+
     public void testFStringParsing() throws ParseException, BadLocationException {
+        checkStr("\\N{foo}", ArrayUtils.asSet(""));
```
Comment (fabioz) [inline plugins/org.python.pydev.parser/tests/org/python/pydev/parser/FStringsParserTest.java:94]: Please add some cases testing the string replacement directly (as mentioned before) and add more tests for the corner cases listed before.

Also, you should be able to call `checkExprs("\\N{foo}", ArrayUtils.asSet());` and it should work (you don't need the `checkStr` -- I think you have become confused because doing `ArrayUtils.asSet()` and `ArrayUtils.asSet("")`, both have the same representation in the error output, even though `assertEquals(ArrayUtils.asSet(), ArrayUtils.asSet(""));` is really expected to fail (because one is really an empty set and the other is a set with a single entry but which doesn't have any representation.

File: plugins/org.python.pydev.parser/src/org/python/pydev/parser/grammar_fstrings/build.xml [hunk 7]
```diff
@@ -1,7 +1,10 @@
 <project name="jython-parser" default="parser">
+	
+	
 
-    <property name="javaccHome" value="c:\bin\javacc-5.0" />
-    <property name="python" value="c:\bin\Miniconda\envs\py27_tests\python.exe" />
+	<property name="env" value="C:\Users\olccjr"/>
```
Comment (fabioz) [inline plugins/org.python.pydev.parser/src/org/python/pydev/parser/grammar_fstrings/build.xml:5]: See: https://coderanch.com/t/108737/build-tools/read-environment-variables-Ant (or google for `ant use environment variable` for more info).

File: plugins/org.python.pydev.parser/src/org/python/pydev/parser/grammar_fstrings/build.xml [hunk 8]
```diff
@@ -1,7 +1,11 @@
 <project name="jython-parser" default="parser">
+	
+	
 
-    <property name="javaccHome" value="c:\bin\javacc-5.0" />
-    <property name="python" value="c:\bin\Miniconda\envs\py27_tests\python.exe" />
+	<property environment="env"/>
+	<property name="userProfile" value="${env.USERPROFILE}"/>
+    <property name="javaccHome" value="${userProfile}\Documents\Javacc\javacc-5.0" />
```
Comment (fabioz) [inline plugins/org.python.pydev.parser/src/org/python/pydev/parser/grammar_fstrings/build.xml:7]: The paths are still hardcoded... the idea is having 2 (made up) variables without anything hardcoded (say, GRAMMAR_JAVACC_DIR and GRAMMAR_PYTHON_EXE).

**Comment by fabioz** [comment]
Just merged ;)
